contrib/rust/src/python/pants/contrib/rust/tasks/cargo_workspace.py

```diff
@@ -0,0 +1,144 @@
+    if target_definition.kind == 'lib' or \
+            target_definition.kind == 'bin' or \
+            target_definition.kind == 'cdylib' or \
+            target_definition.kind == 'rlib' or \
+            target_definition.kind == 'dylib' or \
+            target_definition.kind == 'staticlib' or \
+            target_definition.kind == 'proc-macro':
+      return True
+    else:
+      return False
+
+  @staticmethod
+  def is_test_target(target_definition):
+    if target_definition.kind == 'test' or target_definition.compile_mode == 'test':
+      return True
+    else:
+      return False
+
+  def inject_member_target(self, target_definition, workspace_target):
+
+    def find_member(target_name, members):
+      for member in members:
+        name, _, _ = member
+        if target_name == name:
+          return member
+
+    member_definitions = tuple(
+        (name, path, workspace_target.include_sources)
+        for (name, path) in zip(workspace_target.member_names, workspace_target.member_paths))
+
+    member_definition = find_member(target_definition.name, member_definitions)
+    target_sources = self.get_member_sources_files(member_definition)
+
+    if self.is_test_target(target_definition):
+      synthetic_target_type = self._project_target_kind['test']
+    else:
+      synthetic_target_type = self._project_target_kind[target_definition.kind]
+
+    synthetic_of_original_target = self.context.add_new_target(
+        address=target_definition.address,
+        target_type=synthetic_target_type,
+        derived_from=workspace_target,
+        dependencies=workspace_target.dependencies,
+        cargo_invocation=target_definition.invocation,
+        sources=target_sources)
+
+    self.inject_synthetic_of_original_target_into_build_graph(synthetic_of_original_target,
+                                                              workspace_target)
+
+  def inject_synthetic_of_original_target_into_build_graph(self, synthetic_target, original_target):
+    # https://github.com/pantsbuild/pants/blob/fedc91cb2e1455b7a8dca9c843bbd8b553a04241/src/python/pants/task/simple_codegen_task.py#L359
+    build_graph = self.context.build_graph
+    for dependent_address in build_graph.dependents_of(original_target.address):
+      build_graph.inject_dependency(
+          dependent=dependent_address,
+          dependency=synthetic_target.address,
+      )
+
+    for concrete_dependency_address in build_graph.dependencies_of(original_target.address):
+      build_graph.inject_dependency(
+          dependent=synthetic_target.address,
+          dependency=concrete_dependency_address,
+      )
+
+    if original_target in self.context.target_roots:
+      self.context.target_roots.append(synthetic_target)
+
+  def get_member_sources_files(self, member_definition):
+    _, path, include_sources = member_definition
+    rglobs = RGlobs.to_filespec(include_sources, root=path)
+    path_globs = [PathGlobsAndRoot(
```

Comment:
To capture a `Snapshot` at the buildroot, you should be able to skip the `AndRoot` part here and request something like:
```
snapshot = self.context._scheduler.product_request(Snapshot, [PathGlobs(tuple(rglobs['globs']))])
```

...which has the advantage of being fully cachable.